### PR TITLE
Small refactor in spelling checker

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Refactor prior to #6137

### CODE REVIEW

File: pylint/checkers/spelling.py [hunk 1]
```diff
@@ -260,12 +260,14 @@ class SpellingChecker(BaseTokenChecker):
                 "default": "fmt: on,fmt: off,noqa:,noqa,nosec,isort:skip,mypy:",
                 "type": "string",
                 "metavar": "<comma separated words>",
-                "help": "List of comma separated words that should be considered directives if they appear and the beginning of a comment and should not be checked.",
+                "help": "List of comma separated words that should be considered "
+                "directives if they appear and the beginning of a comment "
```
Comment (DanielNoord) [inline pylint/checkers/spelling.py:264]:
```suggestion
                "directives if they appear at the beginning of a comment "
```

File: pylint/checkers/spelling.py [hunk 2]
```diff
@@ -304,7 +306,7 @@ def open(self):
             self.spelling_dict = enchant.Dict(dict_name)
 
         if self.linter.config.spelling_store_unknown_words:
-            self.unknown_words = set()
+            self.unknown_words: set = set()
```
Comment (DanielNoord) [inline pylint/checkers/spelling.py:309]: Do you know the type of the things in the `set`?

File: pylint/checkers/spelling.py [hunk 3]
```diff
@@ -321,14 +323,17 @@ def open(self):
...
             self.private_dict_file.close()
 
-    def _check_spelling(self, msgid, line, line_num):
+    def _check_spelling(self, msgid: str, line: str, line_num: int) -> None:
         original_line = line
         try:
-            initial_space = re.search(r"^[^\S]\s*", line).regs[0][1]
+            # We choose to ask for forgiveness, mypy does not understand it
+            # The code to check that we match and that there is at least 2 chars
+            # in the first match is way more complex
```
Comment (DanielNoord) [inline pylint/checkers/spelling.py:335]:
```suggestion
            # The mypy warning is caught by the except statement
```

File: pylint/checkers/spelling.py [hunk 4]
```diff
@@ -378,7 +383,10 @@ def _check_spelling(self, msgid, line, line_num):
 
             # Store word to private dict or raise a message.
             if self.linter.config.spelling_store_unknown_words:
-                if lower_cased_word not in self.unknown_words:
+                if (
+                    lower_cased_word not in self.unknown_words
+                    and self.private_dict_file is not None
```
Comment (DanielNoord) [inline pylint/checkers/spelling.py:388]: Isn't this a change in behaviour?

**Review by DanielNoord — CHANGES_REQUESTED** [review]

File: pylint/checkers/spelling.py (hunk 4 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/spelling.py:388]: If ``self.private_dict_file is None`` the check() call would crash pylint. It's never happening but you have to read the whole class workflow to realize that, and mypy can't.  I'll do a follow-up refactor to initialize thing in the constructor where it should be clearer.

File: pylint/checkers/spelling.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/spelling.py:309]: Yes it's strings.
```suggestion
            self.unknown_words: set[str] = set()
```

File: pylint/checkers/spelling.py [hunk 5]
```diff
@@ -321,14 +323,15 @@ def open(self):
...
         if self.private_dict_file:
             self.private_dict_file.close()
 
-    def _check_spelling(self, msgid, line, line_num):
+    def _check_spelling(self, msgid: str, line: str, line_num: int) -> None:
         original_line = line
         try:
-            initial_space = re.search(r"^[^\S]\s*", line).regs[0][1]
+            # The mypy warning is caught by the except statement
+            initial_space = re.search(r"^\s+", line).regs[0][1]  # type: ignore[union-attr]
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/spelling.py:334]: There's a simplification in the regex here on top of the mypy ignore. ``[^\S]`` is ``\s`` and ``\s\s*`` is ``\s+``

**Pull request merged**